lib/Sema/CSApply.cpp

```diff
@@ -2514,6 +2514,65 @@ namespace {
             /*implicit=*/expr->isImplicit(), Type(), getType);
         result = finishApply(apply, Type(), cs.getConstraintLocator(expr));
       }
+			
+      // If the base type is optional and we're referring to an nominal
+      // type member via the dot syntax and the member name matches
+      // Optional<T>.none then emit a diagnostic
+      if (auto baseUnwrappedType = baseTy->lookThroughAllOptionalTypes()) {
+        if (auto DSCE = dyn_cast<DotSyntaxCallExpr>(result)) {
+          
+          Decl *structOrEnumMember = nullptr;
+          StringRef memberName;
+          auto calledValue = DSCE->getCalledValue();
+          bool isOptional = false;
+          
+          if (auto EED = dyn_cast<EnumElementDecl>(calledValue)) {
+            structOrEnumMember = EED;
+            memberName = EED->getNameStr();
+            isOptional = EED->getParentEnum()->isOptionalDecl() &&
+                         !EED->getParentEnum()->getElement(EED->getName())
+                                              ->hasAssociatedValues();
+          } else if (auto VD = dyn_cast<VarDecl>(calledValue)) {
+            if (VD->isStatic()) {
+              structOrEnumMember = VD;
+              memberName = VD->getNameStr();
+              isOptional = VD->getType()->getOptionalObjectType() || nullptr;
+            }
+          }
+          
+          if (structOrEnumMember && isOptional && !memberName.empty()) {
+            auto baseTypeNominal = baseUnwrappedType
+                                   ->getNominalOrBoundGenericNominal();
+            auto memberNameAsIdentifier = cs.getASTContext()
+                                            .getIdentifier(memberName);
+            auto results = tc.lookupMember(baseTypeNominal->getModuleContext(),
+                                           baseUnwrappedType,
+                                           memberNameAsIdentifier,
+                                           defaultMemberLookupOptions);
+            
+            for (LookupResultEntry result : results) {
+              auto value = result.getValueDecl();
+              
+              if (isa<EnumElementDecl>(value) || isa<VarDecl>(value)) {
+                auto baseTypeName = baseUnwrappedType->getString();
+                auto loc = DSCE->getLoc();
+                auto startLoc = DSCE->getStartLoc();
+                
+                tc.diagnose(loc, swift::diag::optional_ambiguous_case_ref,
+                            baseTypeName, memberName);
+                
+                tc.diagnose(loc, swift::diag::optional_fixit_ambiguous_case_ref)
+                .fixItInsert(startLoc, "Optional");
```

Comment:
Nitpick: increase the indent here since its part of the previous statement

Reply:
Done!